dopamine/discrete_domains/run_experiment.py

```diff
@@ -555,10 +570,36 @@ def __init__(
         environment, and returns an agent.
       create_environment_fn: A function which receives a problem name and
         creates a Gym environment for that problem (e.g. an Atari 2600 game).
+      max_simultaneous_iterations: int, maximum number of iterations running
+        simultaneously in separate threads.
       **kwargs: Additional positional arguments.
     """
     threading_utils.initialize_local_attributes(
         self, _environment=create_environment_fn)
+    self._running_iterations = threading.Semaphore(max_simultaneous_iterations)
+    self._output_lock = threading.Lock()
+
     super(AsyncRunner, self).__init__(
         base_dir=base_dir, create_agent_fn=create_agent_fn,
         create_environment_fn=create_environment_fn, **kwargs)
+
+  # TODO(aarg): Decouple experience generation from training.
+  def _run_experiment_loop(self):
+    """Runs iterations in multiple threads until `num_iterations` is reached."""
+    threads = []
+    for iteration in range(self._start_iteration, self._num_iterations):
+      self._running_iterations.acquire()
+      threads.append(self._run_one_iteration(iteration))
+    # Wait for all running iterations to complete.
+    for thread in threads:
+      thread.join()
+
+  @async_method
+  def _run_one_iteration(self, iteration):
+    """Runs one iteration in separate thread, logs and checkpoints results."""
+    statistics = super(AsyncRunner, self)._run_one_iteration(iteration)
+    with self._output_lock:
+      self._log_experiment(iteration, statistics)
+      self._checkpoint_experiment(iteration)
+    tf.logging.info('Completed iteration %d.', iteration)
+    self._running_iterations.release()
```

Comment:
nit: What do you think about moving the release up to `_run_experiment_loop`?  Once `acquire` and `release` are in different methods, it gets harder to track (though it's not so bad here because these functions are so physically close together and logically coupled).

Reply:
the problem is that if we release in `_run_experiment_loop` it means we directly release after starting a non-blocking iteration, which means the semaphore is always empty. Here we want to acquire when we start an iteration, and release when the iteration completes (which can only be tracked in the iteration method).

We could move the `acquire` to the iteration method - but I prefer to have it in the loop because because otherwise the `acquire` can get delay since it would be running in another thread, which would be problematic.

The goal of this is to control the number of thread that start at a time and maintain it below `max_simultaneous_iterations`

Reply:
Understood -- thanks.  I'd missed that the call to append a thread didn't block on its completion.

 Maybe the awkwardness comes from trying to implement [queue](https://docs.python.org/3/library/queue.html).  I think switching from `threading.Semaphore` to `queue` would be low-effort.  Do you think, at this point, that it's worth the investment (or for that matter, a good idea at all)?

Reply:
Thanks for mentioning queues, it seems to be a great candidate to properly handle this kind of tasks. I am not sure though I see the advantages of a queue over semaphores in this specific case.

So to answer your question, I think it's a good idea - but I don't see the gain of switching from the current implementation with semaphore, which would require a little bit of work (to update the way iteration counter works).

I keep the queue idea in mind though and I think this would be great to handle the decoupling of training steps.

Reply:
If I see an opportunity to switch to queues I will. I think queues also work well with the threading_utils local variables are the worker threads would be maintained through the entire experiment.

Adding this as a TODO actually.

Reply:
> I am not sure though I see the advantages of a queue over semaphores in this specific case.

My own preference for queues via `queue` over queue via `Semaphore` is:

- using a standard library over a re-implementation of the standard library makes code more immediately understandable to a reader

- `Semaphore` is a more general utility than 'queue`.  It takes a reader a bit of thought to ensure that `queue`s are the only intended behavior

- using a widely-tested implementation means fewer opportunities for bugs or weird, undocumented interactions in the corner cases

Totally fine to put as a TODO.  I just wanted to unresolve to speak to this specific point.

Reply:
Thanks for the clarification it makes sense. Leaving as a TODO for now